File: codegraph-core/src/main/java/com/dnfeitosa/codegraph/core/models/Artifact.java
/**
 * Copyright (C) 2015-2017 Diego Feitosa [dnfeitosa@gmail.com]
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.dnfeitosa.codegraph.core.models;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static com.dnfeitosa.codegraph.core.utils.Arrays.asList;

public class Artifact implements Comparable<Artifact> {

    private final String id;
    private final String organization;
    private final String name;
    private final Version version;

    private final Set<Dependency> dependencies = new HashSet<>();

    public Artifact(String organization, String name, Version version) {
        this.id = id(organization, name, version);
        this.name = name;
        this.organization = organization;
        this.version = version;
    }

    public String getId() {
        return id;
    }

    public String getOrganization() {
        return organization;
    }

    public String getName() {
        return name;
    }

    public Version getVersion() {
        return version;
    }

    public Artifact addDependency(Dependency dependency, Dependency... dependencies) {
        this.dependencies.add(dependency);
        this.dependencies.addAll(asList(dependencies));
        return this;
    }

    public Set<Dependency> getDependencies() {
        return dependencies;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Artifact artifact = (Artifact) o;
        return Objects.equals(id, artifact.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public int compareTo(Artifact o) {
        return id.compareTo(o.id);
    }

    @Override
    public String toString() {
        return id;
    }

    public static String id(String organization, String name, Version version) {
        return String.format("%s:%s:%s", organization, name, version.getNumber());
    }

    public String hash() {
        return getClass().getName() + "@" + Integer.toHexString(hashCode());
    }
}

File: codegraph-viewer/frontend/app/models/artifact.js
/*
 * Copyright (C) 2015-2018 Diego Feitosa [dnfeitosa@gmail.com]
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
class Artifact {
    constructor(id, name, organization, version, dependencies) {
        this.id = id;
        this.name = name;
        this.organization = organization;
        this.version = version;
        this.dependencies = dependencies || [];
    }

    addDependency(dependency) {
        this.dependencies.push(dependency);
    }
}

class Dependency {
    constructor(artifact, configurations) {
        this.artifact = artifact;
        this.id = artifact.id;
        this.configurations = configurations;
    }
}

class Node {
    constructor(value) {
        this.value = value;
    }

    get id() {
        if (this.value instanceof Artifact) {
            return `${this.value.id}`;
        }
        throw "Not supposed to come here";
    }

    get label() {
        if (this.value instanceof Artifact) {
            return this.value.name;
        }
        if (this.value instanceof Dependency) {
            let artifact = this.value.artifact;
            return `${artifact.organization}:${artifact.name}`;
        }
        return 'n/a';
    }
}


class Edge {
    constructor(source, target) {
        this.source = source.id;
        this.target = target.id;
        this.id = `${source.id}:${target.id}`;
    }
}

class Graph {
    constructor(nodes, edges) {
        this.nodes = nodes;
        this.edges = edges;
    }
}

export {
    Artifact,
    Dependency,
    Node,
    Edge,
    Graph
}
